### PR TITLE
add gumbel_softmax, based on Eric Jang's implementation

### PR DESCRIPTION
**Pull request by hughperkins**
based on:
- https://discuss.pytorch.org/t/stop-gradients-for-st-gumbel-softmax/530/5
- https://github.com/ericjang/gumbel-softmax/blob/3c8584924603869e90ca74ac20a6a03d99a91ef9/Categorical%20VAE.ipynb (section 1)

Gumbel softmax lets you use the reparameterization trick for discrete variables

### CODE REVIEW

**Comment by hughperkins** [comment]
(test output:

```
$ python test/test_nn.py -v -b TestNN.test_gumbel_softmax_st
test_gumbel_softmax_st (__main__.TestNN) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.020s

OK
```
)

File: torch/nn/functional.py
Comment (ozancaglayan) [inline torch/nn/functional.py]: An extra "i" in probabilit(i)y :)

Comment (hughperkins) [reply torch/nn/functional.py]: Good point. Fixed in 80e53d7

**Comment by Kaixhin** [comment]
Thanks for this Hugh, but we may want to hold off on merging this until the [distributions API](http://pytorch.org/docs/master/distributions.html) is more settled - pinging @apaszke to see what the status is.

If we do take this PR, it'll need to be written with the new API, so `torch.distributions.Gumbel` etc.

**Comment by hughperkins** [comment]
Hmmm, leaving aside the `distributions` bit for now, slight nuance on this PR: if we want to calculate entropy ,we need the underlying probabilities, so might be worth refactorizing something like, conceptually:

```
probs = gumbel_softmax(logits)
onehot = gumbel_harden(probs)
```
(instead of:
```
onehot = gumbel_softmax(logits, hard=True)
```
)
Then, the entropy will be simply:
```
eps = 1e-8
probs = probs + eps
entropy = - (probs * probs.log()).sum(1).sum()
```

**Comment by hughperkins** [comment]
note to whoever picks this up (might be me, but probably not until current master is available in versioned form): need to do something like:
```
type_constr = torch.cuda if logits.is_cuda else torch
y_hard = type_constr.FloatTensor( ...
```
(or whatever is the standard approach for cuda-izing things).

**Comment by nhynes** [comment]
> (or whatever is the standard approach for cuda-izing things)

Probably something vaguely similar to
```diff
diff --git a/torch/nn/functional.py b/torch/nn/functional.py
index 81654b2..55b272f 100644
--- a/torch/nn/functional.py
+++ b/torch/nn/functional.py
@@ -764,7 +764,7 @@ def softmax(input, dim=None, _stacklevel=3):
     return torch._C._nn.softmax(input, dim)


-def sample_gumbel(shape, eps=1e-10):
+def sample_gumbel(shape, eps=1e-10, out=None):
     """
     Sample from Gumbel(0, 1)

@@ -772,7 +772,7 @@ def sample_gumbel(shape, eps=1e-10):
     https://github.com/ericjang/gumbel-softmax/blob/3c8584924603869e90ca74ac20a6a03d99a91ef9/Categorical%20VAE.ipynb ,
     (MIT license)
     """
-    U = torch.rand(shape).float()
+    U = out.resize_(shape).uniform_() if out is not None else torch.rand(shape)
     return - torch.log(eps - torch.log(U + eps))


@@ -785,7 +785,7 @@ def gumbel_softmax_sample(logits, tau=1, eps=1e-10):
     (MIT license)
     """
     dims = len(logits.size())
-    gumbel_noise = sample_gumbel(logits.size(), eps=eps)
+    gumbel_noise = sample_gumbel(logits.size(), eps=eps, out=logits.data)
     y = logits + Variable(gumbel_noise)
     return softmax(y / tau, dims - 1)

@@ -816,7 +816,7 @@ def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10):
         _, k = y_soft.data.max(-1)
         # this bit is based on
         # https://discuss.pytorch.org/t/stop-gradients-for-st-gumbel-softmax/530/5
-        y_hard = torch.FloatTensor(*shape).zero_().scatter_(-1, k.view(-1, 1), 1.0)
+        y_hard = logits.data.new(*shape).zero_().scatter_(-1, k.view(-1, 1), 1.0)
         # this cool bit of code achieves two things:
         # - makes the output value exactly one-hot (since we add then
         #   subtract y_soft value)
```

**Comment by hughperkins** [comment]
cool. do you want to PR that onto my branch? or ... ?

**Comment by nhynes** [comment]
GH doesn't seem to like that you renamed pytorch to pytorch-pytorch. Here's a [patch file](https://gist.github.com/nhynes/2669206d6a6d50fccdca0022c4df2016) with tests, if you're up for `git am`ing it manually :)

**Comment by hughperkins** [comment]
> GH doesn't seem to like that you renamed pytorch to pytorch-pytorch

Haha :) Seems like there is already a repo at `hughperkins/pytorch` though, so :P

Added in your commit.

**Comment by nhynes** [comment]
```
FAIL: test_gumbel_softmax_st (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 1021, in test_gumbel_softmax_st
    self._test_gumbel_softmax_st(False)
  File "test_nn.py", line 1005, in _test_gumbel_softmax_st
    assert logits_var.grad.abs().min().data[0] > 0.001
AssertionError
```

This assertion is flaky; I actually had to change the RNG seed to get it to work locally. Perhaps `manual_seed`ing has some machine-specific dependency? Is there a reason for `0.001` specifically or does `> eps` also work?

**Comment by hughperkins** [comment]
We want to show there is actually gradient being backpropagated. If random number generator is not invariant across hardware, then I guess we may as well remove the seed. I'd rather not put `eps`, since `eps` is used for showing things are effectively the same, whereas we want to check that the gradient is effectively not the same, ie not zero.

What are your thoughts on how we can demonstrate that there is a gradient being back-propagated?

Hmmm. Perhaps, we should replace `.min()` with `.mean()` or `.max()`? `.min()` does seem ... fragile.

**Comment by nhynes** [comment]
Ah, I'd meant `eps` in terms of non-zero. `std` seems to work; another approach might be to just check the fraction of entries that have magnitude greater than `eps = 1e-8`.

**Comment by hughperkins** [comment]
Ah, well, I dont really trust eps to show non-zeroness. Floats are only reliable-ish to 6-7 sfs, and eps is 1e-8, way smaller than that.

Checking the number of entries greater than say 0.001 would be ok.

**Review by soumith — CHANGES_REQUESTED** [review]
sample_gumbel and gumbel_softmax_sample need to be prefixed with `_`, to make them private functions (eg. `_sample_gumbel`). After that this PR is good to merge.

File: torch/nn/functional.py
Comment (soumith) [inline torch/nn/functional.py]: dims = logits.dim()

**Comment by hughperkins** [comment]
Hi. Just noticed, you dont need this migrated to the `.distributions` approach?

**Comment by hughperkins** [comment]
Addressed the `_` naming prefix, and the `logits.dim()` change.

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks a lot @hughperkins !

**Pull request merged**
